### ISSUE TITLE
Make jsPlumb work with webpack

### ISSUE THREAD

**Issue by kentcdodds** [sporritt/jsPlumb#314]
I'm packaging this with webpack and I believe that it's because internally, jsPlumb is depending on jsPlumbUtil as a global. Does that sound right?

**Comment by kentcdodds** [comment]
I believe part of the problem is that `this` is not bound to `window` in the webpack context. But jsPlumb should likely not be depending on the global `jsPlumbUtil` anyway.

**Comment by kentcdodds** [comment]
By the way, [this line](https://github.com/sporritt/jsPlumb/blob/8a360c2440dbcfa79466d5ef3bdb4cc230d25568/src/jsPlumb.js#L21) is where it's throwing the error. Because it's in strict mode and it's referencing jsPlumbUtil which is not defined on the global. This should be referencing whatever the `root` is when `jsPlumbUtil` is created and set to it's `this`.

**Comment by kentcdodds** [comment]
FYI, webpack rocks. My problem has been resolved using `script-loader` which runs a script in the same context as adding a script tag would. But I'd say this is still an issue. Just less urgent for me :-)

**Comment by sporritt** [comment]
jsPlumb does depend on the jsPlumbUtil global.  it could be that an explicit `this` needs to be supplied. perhaps you could try

```
_ju = this.jsPlumbUtil
```

and see if it works.

**Comment by kentcdodds** [comment]
That doesn't work because (apparently) the `this` that jsPlumUtil gets attached to is different from the `this` that that line is running in (which is `window`). I'm not sure why that is.

**Comment by sporritt** [comment]
hmmmmm

**Comment by sporritt** [comment]
@kentcdodds i'm calling this done in 1.7.4 dev. if you get a moment to try it out that'd be good.

**Comment by tomsiwik** [comment]
// naive workaround: https://github.com/sporritt/jsPlumb/issues/437
global.Mottle = { consume: null };

// <-- here same error again regarding this issue
var jsPlumb = require('jsplumb');

**Comment by tomsiwik** [comment]
Any updates?

```
jsPlumb-2.1.0.js:2959 Uncaught ReferenceError: jsPlumbUtil is not defined
```

**Comment by sporritt** [comment]
Please supply a test case. 2.1.0 is still in development.

**Comment by tomsiwik** [comment]
Phew... not sure how to do that / where to start. Especially not with qunit. Will eat up some tuts.

**Comment by sporritt** [comment]
ok wait...tell me the exact scenario in which you're having an issue?

**Comment by tomsiwik** [comment]
simply loading jsplumb via `require('jsplumb')` in my backbone app which bundled in webpack

**Comment by sporritt** [comment]
er..yeah.  bundled how in webpack though?

**Comment by tomsiwik** [comment]
Woops. Apologies 'webpacked' was in a different project (same error though). In this one i simply `browserify` the whole app.

```javascript
var b = browserify(bundleConfig).transform(aliasify); //default config & aliasify uses my mvc folder structure
...
return b.bundle();
```

Should i test bundling it from within my 'lib' folder of my mvc folder structure rather than from npm dependancies and check if it still works?

**Comment by tomsiwik** [comment]
Meh... nothing works. Here my basic config:

`config.js`:
```javascript
bundleConfigs: [{
        entries: src + '/javascript/app.js',
        dest: dest + '/js',
        outputName: 'bundle.js',
        extensions: ['.coffee', '.js', '.hbs']
}]
```

`gulptask_browserify.js`:
```javascript
var b = browserify(bundleConfig);

var bundle = function () {
    return b
        .bundle()
        .on('error', handleErrors)
        .pipe(source(bundleConfig.outputName))
        .pipe(gulp.dest(bundleConfig.dest))
        .on('end', reportFinished);
};

...

b.bundle();
```

`app.js`:
```javascript
var jsPlumb = global.jsPlumb = require('jsplumb');
// ERROR HERE: Uncaught ReferenceError: jsPlumbUtil is not defined
// src bundle.js @ jsPlumbUtil.extend(jsPlumbUIComponent, jsPlumbUtil.EventGenerator, {

// here more dependancies that always work... and below that doesn't even get executed...

app = new Marionette.Application({

    initialize: function() {
        this.LineDrawer = jsPlumb.getInstance();
        this.LineDrawer.importDefaults({
          PaintStyle:{
            lineWidth:1,
            strokeStyle:"#FFFFFF"
          },
          Connector:[ "Bezier", { curviness: 30 } ],
          Endpoint:[ "Dot", { radius:4 } ],
          EndpointStyle : { fillStyle: "#FFFFFF"  },
          Anchors : [ "LeftCenter", "RightCenter" ]
        });

        //bla...
    },

...

```

**Comment by sporritt** [comment]
you could try pulling the latest 2.1.0 build and see how that goes. i made a couple of changes. i dont have a great deal of time right now to look at this but i am keen to get it fixed and move on.

**Comment by tomsiwik** [comment]
Nice. Error disappeared, new one appeared :+1: 

`Uncaught ReferenceError: jsPlumbUIComponent is not defined`

**Comment by sporritt** [comment]
ok that's good...um..try again?

**Comment by tomsiwik** [comment]
still jsPlumbUIComponent is not defined

`_ju.extend(_jp.OverlayCapableJsPlumbUIComponent, jsPlumbUIComponent, {`

**Comment by sporritt** [comment]
ah..ahem. yes, forgot to push the built version.

**Comment by tomsiwik** [comment]
Yay... progress: `Uncaught TypeError: Cannot read property 'prototype' of undefined`

`root.jsPlumbInstance.prototype.EndpointDropHandler = function (dhParams) {`

**Comment by sporritt** [comment]
your optimistic patience is appreciated.  once again?

**Comment by tomsiwik** [comment]
Still same error

**Comment by sporritt** [comment]
hmm. maybe i can just try requiring this in node. would that kind of replicate what you're doing, do you think?

**Comment by tomsiwik** [comment]
Well... kinda yeah. But i'm not sure if bundling messes it up somehow.

**Comment by sporritt** [comment]
aha! let's just say bundling messes it up and it's not my fault.  he he he.

**Comment by tomsiwik** [comment]
Exactly xD

**Comment by sporritt** [comment]
sweet. i would mark this as wontfix but it's already closed.

node doesnt like that jsPlumb asks questions about `navigator`. Which is understandable.

**Comment by tomsiwik** [comment]
Soooo... I won't be able to `require('jsplumb')` any time soon?! Don't know the 'wontfix' tag.

**Comment by sporritt** [comment]
wontfix comes from the grandaddy of all issue trackers, bugzilla.  well maybe it isn't considered the grandaddy; who knows. you can get two developers to disagree about what each of their names are without very much effort.  

but this dissembling doesn't help you.  you should be able to require jsplumb. but i will have to replicate what you're doing i suppose.  if you just browserify jsplumb on its own does it fail? presumably it does.

**Comment by tomsiwik** [comment]
Need to test browserifying jsplumb alone, but yeah... it's just ripping of the top and bottom of my bundle.js script. i guess it'll just error out...
Since i'm programming a client-side software with backbone and it's hard to include jsplumb into their model-view pattern i'll just wait for now until it's resolved. Take your time... for me it's not insanely urgent. Thanks for the help so far.

**Comment by tomsiwik** [comment]
Mind looking at it again? It's been a while. I've also added a browserify test-case if it's of any use for you.

**Comment by mdemin914** [comment]
Hey, I am also getting this issue using gulp and browserify. It throws the not defined from a simple require from a node app. 

`var jsPlumb = require('jsplumb');`

The browser throws jsPlumb is not defined from:
`jsPlumb.extend(jsPlumbInstance.prototype, `

I am using jsPlumb 2.0.7, node v0.12.5

Let me know if theres more info that you need.

**Comment by ajainarayanan** [comment]
@sporritt  Is there a project/repo you have that has jsplumb integrated with webpack working? For some reason it either throws `jsPlumbUtil` not found or `jsPlumb.getRenderModes` is not a function errors. As a standalone library with a global variable it works fine. When trying to use it in webpack that is when it is failing with the above errors. Is there any documentation that I am missing?

**Comment by SanthoshHari912** [comment]
@ajainarayanan & @sporritt  I face the same issue while using jsplumb with webpack. I get this error,
"Uncaught TypeError: jsPlumb.getRenderModes is not a function(anonymous function)" in init method.

An example as requested by Ajai will  be really great.

**Comment by ajainarayanan** [comment]
@SanthoshHari912 This is what I tried and I was able to make it work for my project. Not sure if this is the best way but this will hopefully unblock you from using `jsPlumb`.

my `webpack.config.js` looks like this,
```
module: {
    loaders: [
      {
        test: require.resolve('jsPlumb'),
        loaders: [
          'imports?this=>window',
          'script'
        ]
      },
      { ...}
...
]
```
Now in your app, in any specific file you could just do `require('jsPlumb')` and `jsPlumb` will be available as a variable within that module(file). I am still on the early stages of working on my project with `jsPlumb` + `webpack` but I will update this thread once I publish my repo in github as an example.

**Comment by edi9999** [comment]
Hi @ajainarayanan , I'm encountering thte same issue, but I don't think your code snippet is correct, it should rather be require.resolve('jsplumb'). 

Also, it requires you to install `npm install imports-loader script-loader --save`

**Comment by edi9999** [comment]
It would still be great to have jsPlumb outside of the window object, it is usually not recommended to pollute the global window object

**Comment by ajainarayanan** [comment]
@edi9999 Hhhmm Not sure why its working for me but `require.resolve('jsPlumb')` seems to be working for me. As far as I remember when you use `jsPlumb` as a standalone script the object that gets added to the `window` object is `jsPlumb` so thats my assumption to use `require.resolve('jsPlumb)`.

Regarding polluting window object I don't think its polluting global object. It just gets injected in the context of that file that I am require'ing it. If try to access it in any other context/module I will get a Reference Error as it is not injected into that particular context. So no `require('jsPlumb')` doesn't get injected to `window` object ( as far as I have seen. Feel free to correct me if its wrong).

Regarding `imports` & `script` loader, yes obviously they are loaders and have to installed. I kind of assumed for anyone who is using webpack to recognize it as loaders and should install them.

**Comment by SanthoshHari912** [comment]
@ajainarayanan and @edi9999 

Thanks for your suggestions. I don't get this error for 'getRenderModes' but at same time jsPlumb.default object is empty now. I get the below error,
**Uncaught TypeError: _jsPlumb2.default.ready is not a function**

**Comment by edi9999** [comment]
> require.resolve('jsPlumb') seems to be working for me

Are you running on windows ? The filesystem is case insensitive there, maybe you will have issues when running on unix later.

> So no require('jsPlumb') doesn't get injected to window object ( as far as I have seen. Feel free to correct me if its wrong).

That's why I'm experiencing. 

If I do : 

```
require('jsplumb');
const jsPlumb = window.jsPlumb;
console.log(jsPlumb)
```

I have a local reference, but a global reference is still leaked.

**Comment by edi9999** [comment]
For example, if I run window.jsPlumb from the browser console, I get the global jsPlumb object

**Comment by sporritt** [comment]
i wouldn't bother discussing removing jsPlumb from the window object on here, because it isn't going to happen. 

let's instead focus on what people are asking for.  i see a bit of `require` going on and the occasional webpack config but I wouldn't say that thus far anyone has provided full details of their setup.  or, indeed, what their actual goal is, other than "make it work with webpack".  maybe, @edi9999, you're the guy who can provide a little more clarity.

**Comment by sporritt** [comment]
there may be some pertinent information in issue #399.

**Comment by edi9999** [comment]
The goal is that people using JS loaders (such as webpack, but also browserify), can use the library without having to install multiple plugins. 

One way to do this is to **not** minify the code yourself, but instead use the "node.js" require to call other modules. Module loaders parse the code that is required and returns whatever `module.exports` contains (which exports some part of the code) . In your case, module.exports contains nothing, so when you do `var p = require("jsplumb")`, you get a null object. 

If you just added `if(module) { module.exports = jsPlumb } ` in your dist/js/jsplumb.min.js file, it would work seamlessly with weppack and browserify.

**Comment by sporritt** [comment]
like this, you mean?

https://github.com/sporritt/jsPlumb/blob/master/src/jsPlumb.js#L3128

**Comment by sporritt** [comment]
what i was hoping for was an example setup i could use to put an example into jsplumb. would it be sufficient for everyone if i added a demo with a package json requiring webpack and jsplumb and then a test page?

**Comment by 4leem** [comment]
Yes indeed, it will be helpful if we have a demo with webpack + package.json.

**Comment by sporritt** [comment]
@4leem in your use case, do you need to instantiate jsPlumb via the window object:

```
<script>
  jsPlumb.ready(function() {
   .. do your things
  });
</script>
```

or are you looking to require jsPlumb inside some other script:

```
var jsp = require("jsPlumb");
jsp.ready(function() {
    .. do your stuff
});
```

**Comment by sporritt** [comment]
in the 2.2.0 dev branch there's a webpack test setup:

https://github.com/sporritt/jsPlumb/tree/2.2.0/demo/webpack

2.2.0 is not released yet. maybe some of you can pull that branch and try out the page and let me know your thoughts.

**Comment by 4leem** [comment]
Thanks for sharing the demo with webpack. I'm using ES6, `import jsPlumb from 'jsPlumb'`

**Comment by sporritt** [comment]
does it work for you?

**Comment by 4leem** [comment]
Yes, it did work! I'm able to run jsPlumb(v2.2.0) with webpack without additional loaders.

**Comment by sporritt** [comment]
that's good. that's with `import jsPlumb from 'jsPlumb'` ?

**Comment by ajainarayanan** [comment]
###### TL;DR - 
`require('jsPlumb')` works fine if I use `require.resolve('jsplumb')` in my webpack loader but not with the minified file.

###### Detailed version
@sporritt The demo directly requires jsplumb from the `node_modules` folder. It would be nice if we can use `jsplumb` with require or import using the minified version of the library that is shipped. I am referring to something like this,

```
//webpack.config.js

var webpack = require('webpack');
var path = require('path');
var LodashModuleReplacementPlugin = require('lodash-webpack-plugin');

module.exports = {
  context: __dirname,
  entry: {
    'dag': './dag.js',
    'vendor': ['react', 'react-dom', 'redux', 'lodash', 'classname']
  },
  module: {
    loaders: [
      {
        //test: require.resolve('jsPlumb'),
        test: /node_modules\/jsplumb\/dist\/jsPlumb-2.2.0-min.js/,
        loaders: [
          'imports?this=>window',
          'script'
        ]
      },
      {
        test:  /\.(ttf|eot|svg|woff|woff(2))(\?v=[0-9]\.[0-9]\.[0-9])?$/,
        loader: "file-loader"
      },
      {
        test: [
          /\.less$/
        ],
        loader: 'style-loader!css-loader!less-loader'
      },
      {
        test: /\.js$/,
        loader: 'babel',
        exclude: /node_modules/,
        query: {
          plugins: ['lodash'],
          presets: ['react', 'es2015']
        }
      }
    ]
  },
  output: {
    filename: './[name].js',
    path: __dirname + '/dist',
    library: 'dag',
    libraryTarget: 'umd'
  },
  externals: {
    'react': {
      root: 'React',
      commonjs2: 'react',
      commonjs: 'react',
      amd: 'react'
    },
    'react-dom': {
      root: 'ReactDOM',
      commonjs2: 'react-dom',
      commonjs: 'react-dom',
      amd: 'react-dom'
    },
    'classname': 'classname',
    'redux': 'redux',
    'lodash': 'lodash'
  },
  plugins: [
    new LodashModuleReplacementPlugin
  ]
};

```

and in my `app.js`

```
var jsPlumb = require('jsplumb');
// or 
import jsPlumb from 'jsPlumb';
```

I am trying to use the `jsPlumb-2.2.0-min.js` file (which is 5x smaller than the regular file `jsPlumb-2.2.0.js`) and I am getting the same error,

<img width="1279" alt="screen shot 2016-06-15 at 9 22 35 pm" src="https://cloud.githubusercontent.com/assets/1452845/16105433/501012e2-333f-11e6-8795-d476b56648cb.png">

Hope I gave some clarity on what I am trying to do. 

PS: 
1. The `require.resolve('jsplumb')` does work and I will push my project to github as an example.
2. I am ok with adding `jsPlumb` to `window` object but it would be nice if we don't or have a separate dist file that just adds it to the global object and the other that is used in the module system.

**Comment by sporritt** [comment]
you've seen the demo though, right? no `require.resolve('jsplumb')` is necessary.

**Comment by ajainarayanan** [comment]
@sporritt So would this still work? `require('jsPlumb')`? instead of `require('<node_modules_path>/jsPlumb')` in my `app.js`?

**Comment by ajainarayanan** [comment]
@sporritt Sorry about my previous comments. Yes. it works! Thanks a lot!

**Comment by sporritt** [comment]
great!

**Comment by cgarethc** [comment]
This problem seems to be back in version 2.3.0. I can use webpack just fine with 2.2.10, but as soon as it upgrades to 2.3.0 I get:
Uncaught TypeError: _jsplumb2.default.getInstance is not a function
    at new Canvas (bundle.js:47757)
    at bundle.js:68955
    at measureLifeCyclePerf (bundle.js:68735)
    at ReactCompositeComponentWrapper._constructComponentWithoutOwner (bundle.js:68954)

**Comment by sporritt** [comment]
you've seen that there is a webpack demo in jsplumb now?

https://github.com/jsplumb/jsPlumb/tree/master/demo/webpack

**Comment by samiph** [comment]
@cgarethc `import {jsPlumb} from 'jsplumb'` seems to be working with versions > 2.2.10. Tested with 2.5.1. I got the same error as you with `import jsPlumb from 'jsplumb'`
